### PR TITLE
leave the comment in parentheses of argumentless Fn

### PR DESCRIPTION
**Pull request by rchaser53**
related: #3508

### CODE REVIEW

**Comment by rchaser53** [comment]
I didn't implement the logic for the below case.

```
fn foo(
    bar: impl PartialEq<
        // this comment is deleted
    >
) {}
```

Because rustfmt can format and leave the comment for below code.

```
// same meaning above code.
fn foo(
    bar: impl PartialEq,
    // this comment is deleted
) {
}
```

File: src/types.rs [hunk 1]
```diff
@@ -314,46 +316,76 @@ where
...
-        ")",
-        ",",
-        |arg| arg.span().lo(),
-        |arg| arg.span().hi(),
-        |arg| arg.rewrite(context, list_shape),
-        list_lo,
-        span.hi(),
-        false,
-    );
+    let (list_str, tactic) = if inputs.len() == 0 {
```
Comment (scampi) [inline src/types.rs]:
```suggestion
    let (list_str, tactic) = if inputs.is_empty() {
```

File: src/types.rs [hunk 2]
```diff
@@ -314,46 +316,76 @@ where
...
+        let tactic = if output.contains('\n') {
+            DefinitiveListTactic::Vertical
+        } else {
+            definitive_tactic(
+                &Vec::<ListItem>::new(),
+                ListTactic::HorizontalVertical,
+                Separator::Comma,
+                shape.width.saturating_sub(2 + output.len()),
```
Comment (scampi) [inline src/types.rs]: What does `2` represent ? Can you add a comment that describes its meaning ?
I've seen the following format: for example `// 1 = "{"` which means the digit 1 is about the opening curly bracket.

File: src/types.rs [hunk 3]
```diff
@@ -314,46 +316,76 @@ where
...
+            )
+        };
+        let list_hi = context.snippet_provider.span_after_last(span, ")");
```
Comment (scampi) [inline src/types.rs]: Maybe you can add a new method in `source_map::SpanUtils` called `span_before_last`, this way the following could be simplified and avoid the `BytePos(1)` substraction.

File: src/types.rs (hunk 1 above)
Comment (rchaser53) [reply src/types.rs]: Actually I also want to use is_empty, but the below error emits in this case.

```
error[E0658]: use of unstable library feature 'exact_size_is_empty' (see issue #35428)
   --> src/types.rs:321:40
    |
321 |     let (list_str, tactic) = if inputs.is_empty() {
    |                                        ^^^^^^^^
    |
    = help: add #![feature(exact_size_is_empty)] to the crate attributes to enable
```

I can fix easily this. just need to add ```add #![feature(exact_size_is_empty)].```
Can I use a feature attribute in rustfmt?

**Comment by rchaser53** [comment]
Thank you the review. I changed my source.

File: tests/target/issue-3508.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+fn foo<F>(foo2: F)
+where
+    F: Fn(
+       // this comment is deleted
+       ),
```
Comment (topecongiro) [inline tests/target/issue-3508.rs]: The closing paren is indented one level deeper:

```rust
    F: Fn(
       // this comment is deleted
    ),
```

File: tests/target/issue-3508.rs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+fn foo_block<F>(foo2: F)
+where
+    F: Fn(/* this comment is deleted */),
+{
+}
+
+fn bar(
+    bar2: impl Fn(
+    // this comment is deleted
```
Comment (topecongiro) [inline tests/target/issue-3508.rs]: The comment should be indented one level deeper:

```rust
fn bar(
    bar2: impl Fn(
        // this comment is deleted
    ),
```

**Comment by rchaser53** [comment]
Thanks. I fixed about indents.

File: src/types.rs (hunk 1 above)
Comment (rchaser53) [reply src/types.rs]: What I'm saying is [this](https://github.com/rust-lang/rustfmt/pull/3518/commits/b7179d6b9d279e1accae47c44e50ce5e2db5cfd3)

Comment (topecongiro) [reply src/types.rs]: Sorry, I missed the comment the last time I reviewed this PR. I would rather avoid using an unstable feature in rustfmt, as it prevents updating rustfmt in the stable channel. I am sorry but could you revert the https://github.com/rust-lang/rustfmt/pull/3518/commits/b7179d6b9d279e1accae47c44e50ce5e2db5cfd3?

Comment (rchaser53) [reply src/types.rs]: Sure. I rebased it.

File: src/types.rs [hunk 6]
```diff
@@ -381,6 +392,22 @@ fn type_bound_colon(context: &RewriteContext<'_>) -> &'static str {
...
 }
 
+// If the return type is multi-lined, then force to use multiple lines for
+// arguments as well.
+fn get_tactics(item_vec: &Vec<ListItem>, output: &str, shape: &Shape) -> DefinitiveListTactic {
+    if output.contains('\n') {
+        DefinitiveListTactic::Vertical
+    } else {
+        definitive_tactic(
+            &*item_vec,
```
Comment (topecongiro) [inline src/types.rs]: This could be simplified with the above change:
```suggestion
            item_vec,
```

**Comment by rchaser53** [comment]
I improved it.

**Comment by topecongiro** [comment]
Thank you!

**Pull request merged**